src/sidebar/components/thread-list-omega.js

```diff
@@ -0,0 +1,246 @@
+ */
+function ThreadListOmega({ thread, $rootScope }) {
+  const clearSelection = useStore(store => store.clearSelection);
+
+  // Height of the scrollable container. For the moment, this is the same as
+  // the window height.
+  const [windowHeight, setWindowHeight] = useState(window.innerHeight);
+
+  // Scroll offset of scroll container. This is updated after the scroll
+  // container is scrolled, with debouncing to limit update frequency.
+  const [scrollPosition, setScrollPosition] = useState(
+    getScrollContainer().scrollTop
+  );
+
+  // Map of thread ID to measured height of thread.
+  const [threadHeights, setThreadHeights] = useState({});
+
+  // ID of thread to scroll to after the next render. If the thread is not
+  // present, the value persists until it can be "consumed".
+  const [scrollToId, setScrollToId] = useState(null);
+
+  // Effect to scroll to a new annotation when it is created.
+  useEffect(() => {
+    // When a new annotation is created in the application, scroll to that
+    // annotation's thread (unless it is a highlight or reply)
+    $rootScope.$on(events.BEFORE_ANNOTATION_CREATED, (event, annotation) => {
+      if (!isHighlight(annotation) && !isReply(annotation)) {
+        clearSelection();
+
+        // If the annnotation is already visible, scroll directly to it, otherwise
+        // we'll need to calculate its _expected_ position and scroll there.
+        setScrollToId(annotation.$tag);
+      }
+    });
+  }, [$rootScope, clearSelection]);
+
+  // Effect to trigger recalculation of visible threads when scroll
+  // container is scrolled or window is resized.
+  useEffect(() => {
+    const scrollContainer = getScrollContainer();
+
+    const updateScrollPosition = debounce(
+      () => {
+        setWindowHeight(window.innerHeight);
+        setScrollPosition(scrollContainer.scrollTop);
+      },
+      10,
+      { maxWait: 100 }
+    );
+
+    scrollContainer.addEventListener('scroll', updateScrollPosition);
+    window.addEventListener('resize', updateScrollPosition);
+
+    return () => {
+      scrollContainer.removeEventListener('scroll', updateScrollPosition);
+      window.removeEventListener('resize', updateScrollPosition);
+    };
+  }, []);
+
+  const topLevelThreads = thread.children;
+
+  // TODO - We could optimize here by memo-ing such that `topLevelThreadIds`
+  // only updates if the IDs did change.
+  const topLevelThreadIds = topLevelThreads.map(t => t.id);
+
+  // Effect to update measured thread heights when the set of threads
+  // changes.
+  useEffect(() => {
+    let newHeights = {};
+
+    for (let id of topLevelThreadIds) {
+      const threadElement = document.getElementById(id);
+      if (!threadElement) {
+        // Thread is currently off screen.
+        continue;
+      }
+      const height = getElementHeightWithMargins(threadElement);
+      if (height !== null) {
+        newHeights[id] = height;
+      }
+    }
+
+    setThreadHeights(heights => {
```

Comment:
Note that you can pass a function to a state setter. It receives the current state and returns the updated state. This can be useful to avoid the whole effect depending on the earlier state.